packages/ckeditor5-dev-license-checker/src/validate-license-files.ts

```diff
@@ -0,0 +1,347 @@
+/**
+ * @license Copyright (c) 2003-2025, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+import { glob, readFile, writeFile } from 'fs/promises';
+import { findPackageJSON } from 'module';
+import { createPatch } from 'diff';
+import upath from 'upath';
+
+type CopyrightOverride = {
+	packageName: string;
+	dependencies: Array<{
+		license: string;
+		name: string;
+		copyright: string;
+	}>;
+};
+
+type DependencyMapItem = {
+	packageName: string;
+	packagePath: string;
+	dependencies: Array<{
+		name: string;
+		license?: string;
+		copyright?: string;
+	}>;
+};
+
+type ValidationItem = {
+	licensePath: string;
+	licenseMissing?: boolean;
+	sectionMissing?: boolean;
+	updateNeeded?: boolean;
+	updated?: boolean;
+	patch?: string;
+};
+
+const conjunctionFormatter = new Intl.ListFormat( 'en', { style: 'long', type: 'conjunction' } );
+
+/**
+ * @param options
+ * @param options.fix Whether to fix license files instead of printing errors.
+ * @param options.shouldProcessRoot Whether validation should process the root.
+ * @param options.shouldProcessPackages Whether validation should process `packages/*`.
+ * @param options.isPublic Whether license should use disclaimer meant for open source repositories.
+ * @param options.rootDir Base directory.
+ * @param options.projectName Project name referred to in the licenses.
+ * @param options.mainPackageName Designated package that contains collected licenses from all other packages.
+ * @param options.copyrightOverrides Map of of copyright that can both add new ones, as well as override existing ones.
+ *
+ * @returns Exit code of the script that indicates whether it passed or errored.
+ */
+export async function validateLicenseFiles( {
```

Comment:
This function is almost 220 lines long. While it's not a blocker, I'd suggest extracting some logic from it to make it easier to reason about.